Openmp merge

cjllorente827 opened the pull request:
---
name: Pull request
about: Create a pull request to help us improve
title: ''
labels: ''
assignees: ''

---

Thank you so much for your PR! To help us review, fill out the form to the best of your ability.  Please make use of the development guide at https://enzo.readthedocs.io/en/latest/developer_guide/ModificationIntro.html#how-to-develop-enzo

**Pull request summary**

Provide a general summary of your changes in the title above, for example "Fixes crash in FluxCorrection between AMR levels".  Please avoid non-descriptive titles such as "Addresses issue #42."

If you are able to do so, please do not create the PR out of main, but out of a separate branch.

**Describe your pull request**

Please provide at least 1-2 sentences describing the pull request in detail.  Why is this change required?  What problem does it solve?

If it fixes an open issue, please link to the issue here.  Github will [autolink](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/autolinked-references-and-urls) if prefixed with a hash, e.g. #42.

If your PR depends on another PR, please link to the PR here [github autolinking syntax](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/autolinked-references-and-urls).

[ ] This is a major change or addition.  Will require two reviewers.

**PR Checklist**

Note that some of these check boxes may not apply to all pull requests.

- [ ] New features are documented with narrative docs.
- [ ] Adds a test for any bugs fixed. Adds tests for new features.

**Thank you!**

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or the
recommended next step seems overly demanding, or if you would like help in
addressing a reviewer's comments.  Lastly, please ping us if you've been waiting
too long to hear back on your PR.

Commit:
No "fuzzy boundaries" for RT load balancing.

--HG--
branch : week-of-code

Commit:
No "fuzzy boundaries" for RT load balancing.

--HG--
branch : week-of-code

Commit:
Changing 2 defaults.
- SubgridSizeAutoAdjust to TRUE (up to 30% increases in speed reported).
- Optimized CommunicationTransferParticles ON.  Was bracketed by the
OPTIMIZED_CTP define before.

--HG--
branch : week-of-code

Commit:
Empty merge.  I have no idea what I did to result in this.

--HG--
branch : week-of-code

Commit:
If running with only Pop III SF, don't create the Metallicity field.
Placing the HD cooling curve fit ceiling lower at T ~ 1e5K to avoid
overflows (inf) in the tables at T > 1e6K.

--HG--
branch : week-of-code

brittonsmith committed:
Updated kraken compile macro for new lib locations.

--HG--
branch : week-of-code

brittonsmith committed:
Merged.

--HG--
branch : week-of-code

Commit:
Forgot to remove OPTIMIZED_CTP #ifdefs and #ifndefs as this was
changed to the default behavior.

--HG--
branch : week-of-code

bwoshea committed:
fixed some constants for this test problem (thanks to Jun-Hwan Choi)

--HG--
branch : week-of-code

bwoshea committed:
Added a test problem for a sedov explosion in a stratified medium.

--HG--
branch : week-of-code

Commit:
Changing float->FLOAT to agree with grid::PrepareGrid

--HG--
branch : week-of-code

Commit:
Minor optimizations to ray merging.  The radius was a float instead of
FLOAT.  Maybe this was causing some "misplaced" photons when ray
merging is used?

--HG--
branch : week-of-code

Commit:
First attempt at calculating the Lyman-Werner flux using the binary
tree built for ray merging.  Not working.

--HG--
branch : week-of-code

Commit:
First attempt at calculating the Lyman-Werner flux using the binary
tree built for ray merging.  Not working.

--HG--
branch : week-of-code

Commit:
Added basic configuration for OpenMP.  No threading yet.

--HG--
branch : litemp

Commit:
Threading sections that doesn't depend involve communication.

--HG--
branch : litemp

Commit:
Adding C++ versions for inteuler() and intvar() to avoid memory
corruption.  No idea why this happens, but it fixed the problem in the
open-mp branch.

--HG--
branch : litemp

Commit:
Ported improved euler changes to litemp version.  No support for
conservative or positive reconstruction yet.

--HG--
branch : litemp

Commit:
Threading FFT wrapper and potential solver (partially).

--HG--
branch : litemp

Commit:
Bringing in latest changes from week-of-code into litemp.

--HG--
branch : litemp

Commit:
Threading most things except for the ray tracing in EvolvePhotons.
Needed to change ijk (MAX_ANY_SINGLE_DIRECTION) to the maximum
dimension in the grid.  For some reason, it was causing memory
corruption issues in the rate solver.

--HG--
branch : litemp

Commit:
Finally threaded the ray tracing!!

--HG--
branch : litemp

Commit:
Make the default creation time just before the start time.

--HG--
branch : litemp

Commit:
Don't thread the ray tracing if there are less photon packages than cores.

--HG--
branch : litemp

Commit:
Fixing a bug in the ray merging with threading.

--HG--
branch : litemp

Commit:
Grid partitioning works for pure OpenMP runs now (use-mpi-no).
Fortran didn't like multiple threads writing to the same file,
e.g. write(4,*).  Ifdef'ed them out if OpenMP is on.

--HG--
branch : litemp

Commit:
Making the stdout consistent with the hybrid code.

--HG--
branch : litemp

Commit:
Removed negative color warnings temporarily

--HG--
branch : litemp

Commit:
Making the subgrid extrema parameters depend on number of cores
instead of MPI processes.  Removed some minor memory leaks.

--HG--
branch : litemp

Commit:
Merging in latest changes from WoC.

--HG--
branch : litemp

Commit:
SubgridFluxStorage pointer set to NULL when the grid created.

--HG--
branch : litemp

Commit:
Merging.

--HG--
branch : litemp

Commit:
Reducing line length to 72c

--HG--
branch : litemp

Commit:
Merging latest changes from WoC into litemp

--HG--
branch : litemp

Commit:
Making beamed radiation compatible with threading.

--HG--
branch : litemp

Commit:
Converting solve_rate_cool to F90.  All of the temporary (shared)
variables were overflowing the stack with OpenMP.  Turned them into
allocatable variables, which don't have this problem.

--HG--
branch : litemp

Commit:
CurrentTime needs to be adjust when merging rays.  Moved speed of
light conversion to code units to TransportPhotons.

--HG--
branch : litemp

Commit:
Check if the run is parallel before reverting the RT load balancing.

--HG--
branch : litemp

Commit:
Changing grid loop in EvolveLevel to guide scheduling.  Reducing size
of temp. variables to idim instead of ijkn (3072).

--HG--
branch : litemp

Commit:
Forgot about these large temp. arrays.

--HG--
branch : litemp

Commit:
Changing some more loops to guided scheduling.  More threading in
RebuildHierarchy.

--HG--
branch : litemp

Commit:
Deleting allocated GammaField.

--HG--
branch : litemp

Commit:
Merging latest changes in WoC into hybrid version.

--HG--
branch : litemp

Commit:
Turn off "reset load balancing" if running serially.

--HG--
branch : litemp

Commit:
Merging in week-of-code changes.

--HG--
branch : litemp

Commit:
Merging week-of-code changes into litemp

--HG--
branch : litemp

Commit:
Changed scheduling to guided for LW radiation

--HG--
branch : litemp

Commit:
Output position and velocity of the accreting star particle.

--HG--
branch : litemp

Commit:
Changing some debugging positions to GOUTSYM. Allowing for small
subgrids for better load balancing at high levels

--HG--
branch : litemp

jwise77 committed:
Merging in latest changes from week-of-code into litemp.  Compiles.
There were some conflicts in the cooling modules because in the litemp
version, I had to convert solve_rate_cool to F90.  I need to
double-check the merge before pushing.

--HG--
branch : litemp

jwise77 committed:
Merging fix into litemp.  Found more errors in my port of the dust
cooling to the F90 code in the litemp branch.  Still calculates the
dust temperature wrong.

--HG--
branch : litemp

jwise77 committed:
Fixed some precision issues with the new dust cooling routines.

--HG--
branch : litemp

jwise77 committed:
Giving some tolerance for the hypernova and pair-instability SN mass
ranges.

--HG--
branch : litemp

jwise77 committed:
Precision changes for ray merging.

--HG--
branch : week-of-code

jwise77 committed:
Precision changes for ray merging.

--HG--
branch : week-of-code

jwise77 committed:
Precision changes for ray merging.
(transplanted from 0a39894ee43fdded9c892ea34a9c61e195b98d85)

--HG--
branch : litemp

jwise77 committed:
Merging WoC changes into litemp

--HG--
branch : litemp

jwise77 committed:
Adding a command line option to write the dust temperature field to
the dataset.

--HG--
branch : week-of-code

jwise77 committed:
Adding a command line option to write the dust temperature field to
the dataset.

--HG--
branch : week-of-code

jwise77 committed:
Merging

--HG--
branch : week-of-code

jwise77 committed:
Deleting LevelArray linked list.

--HG--
branch : week-of-code

jwise77 committed:
Deleting the LevelArray linked list in the non-grouped datasets.
Setting the default hierarchy output back to HDF5 and ASCII.

--HG--
branch : week-of-code

jwise77 committed:
Fixing a bug that removed gas from the wrong cell after BH accretion.
Also removing density from any existing colour field.

--HG--
branch : week-of-code

jwise77 committed:
First pass at threading the interpolation.

--HG--
branch : litemp

jwise77 committed:
PPM Precision fixes in the C++ translation.

--HG--
branch : litemp

jwise77 committed:
Changing SolveForPotential to guided scheduling

--HG--
branch : litemp

jwise77 committed:
Merging latest changes from week-of-code.

--HG--
branch : litemp

jwise77 committed:
Merging in latest changes with litemp.

--HG--
branch : litemp

jwise77 committed:
Trying to thread some more of RebuildHierarchy.

--HG--
branch : litemp

jwise77 committed:
Fixed recently introduced bug in threading that caused bad MPI sends.

--HG--
branch : litemp

jwise77 committed:
More threading.  CIC depositing and all fortran utilities.

--HG--
branch : litemp

jwise77 committed:
Removed debugging statements.

--HG--
branch : litemp

jwise77 committed:
Merging.

--HG--
branch : week-of-code

jwise77 committed:
Making my old projection routines to have the fields weighted by
density instead of density^2.

--HG--
branch : week-of-code

jwise77 committed:
Consolidating an MPI call.  Avoiding an Allreduce if star_maker9 is
not used.

--HG--
branch : week-of-code

jwise77 committed:
Consolidating an MPI call.  Avoiding an Allreduce if star_maker9 is
not used.

--HG--
branch : litemp

jwise77 committed:
First stab at load balancing on real compute time instead of number of
cells.

--HG--
branch : litemp

jwise77 committed:
Adding more load balancing timers.

--HG--
branch : litemp

jwise77 committed:
Implementing a "fading memory" filter for estimating compute time for
the next timestep.

--HG--
branch : litemp

jwise77 committed:
Debugging new load balancing.  Working pretty good now on small runs.

--HG--
branch : litemp

jwise77 committed:
More debugging of CPU-time load balancing, especially during startup
when there are no timing data.

--HG--
branch : litemp

jwise77 committed:
Optimizations in feedback sphere checking.

--HG--
branch : litemp

Commit:
More optimizations for feedback sphere for better scalability.

--HG--
branch : litemp

jwise77 committed:
Merging in LW (incomplete) tree calculation.  Saved from April 2011.

--HG--
branch : week-of-code

jwise77 committed:
Lyman-Werner radiation calculation now done in a binary tree!

--HG--
branch : week-of-code

jwise77 committed:
Lyman-Werner radiation calculation now done in a binary tree!

--HG--
branch : week-of-code

jwise77 committed:
Creating a wrapper routine to choose the tree calculation if there is
ray merging and >10 sources.

--HG--
branch : week-of-code

jwise77 committed:
Creating a wrapper routine to choose the tree calculation if there is
ray merging and >10 sources.

--HG--
branch : week-of-code

jwise77 committed:
LW tree calculation now works with star particles; optimized the tree
search; debugging

--HG--
branch : week-of-code

jwise77 committed:
LW tree calculation now works with star particles; optimized the tree
search; debugging

--HG--
branch : week-of-code

Commit:
* Move up to 10k grids in one pass in load balancing.
* Load balancing only done on ROOT then broadcast.
* Fixed minor syntax errors where Pathscale complained
* Increase number of send buffers to 200k

--HG--
branch : litemp

jwise77 committed:
Merging latest changes from week-of-code into litemp.

--HG--
branch : litemp

jwise77 committed:
During a merge, I forgot to make this change.

--HG--
branch : litemp

jwise77 committed:
Threading grid::CommunicationSendParticles.  Useful in load balancing.

--HG--
branch : litemp

Commit:
Don't create a radiation source if there's no luminosity.

--HG--
branch : litemp

jwise77 committed:
Cleaning up logic in RT coupled solver.  Also, compute the LW
radiation all at once.

--HG--
branch : litemp

jwise77 committed:
Threading moving subgrid particles from list to arrays.

--HG--
branch : litemp

jwise77 committed:
Threaded moving particles from grid to move list.  This wasn't trivial
as a single #pragma.  Turned off cpu affinity for macs.

--HG--
branch : litemp

jwise77 committed:
Load balance in two stages (fields then particles) to reduce
concurrent MPI calls, like PrepareGravitatingMassField.  Should be
beneficial in high (>1k) MPI counts.  If useful, I will adapt it for
the other load balancers.

--HG--
branch : litemp

jwise77 committed:
Writing observed cost to file for restarting.

--HG--
branch : litemp

jwise77 committed:
Reducing the number of MPI calls during cosmology initializations.
Updated some calls in RK2.  Removing star lifetime limiter on the
timestep.

--HG--
branch : week-of-code

jwise77 committed:
Level loop in RH over all levels and then check if grids exist.

--HG--
branch : week-of-code

jwise77 committed:
Setting the upper limit for Type II SN to 40.1, so 40 (plus a little)
Msun is included.

--HG--
branch : week-of-code

jwise77 committed:
When MultiSpecies is on, the colors are considered as fractional
species.  So we need to advect them as well in the hydro_rk solvers.

--HG--
branch : week-of-code

jwise77 committed:
Adding cell flaggin method to refine by metal mass.

--HG--
branch : week-of-code

jwise77 committed:
Turn off star lifetime timestep limiter

--HG--
branch : week-of-code

jwise77 committed:
Merging.

--HG--
branch : week-of-code

jwise77 committed:
10% speedup to MHD by reordering some loops

--HG--
branch : week-of-code

jwise77 committed:
Merging

--HG--
branch : week-of-code

bwoshea committed:
updating grid stuff - John's suggestion.

--HG--
branch : week-of-code

bwoshea committed:
set up physical density threshold for star maker 2.

--HG--
branch : week-of-code

bwoshea committed:
replaced some physical constants with defines in phys_constants.h (including a couple that were already there).

--HG--
branch : week-of-code

bwoshea committed:
fixed omega_matter/omega_baryon kludge, because it's wrong.

--HG--
branch : week-of-code

bwoshea committed:
Fixed a couple of exponentials so Enzo compiles with Cray compilers.

--HG--
branch : week-of-code

bwoshea committed:
Manisha's blue waters makefiles for Cray and PGI compilers.  Unchecked!  BEWARE!

--HG--
branch : week-of-code

bwoshea committed:
merged.

--HG--
branch : week-of-code

bwoshea committed:
updated Grid_StarParticleHandler.C

--HG--
branch : week-of-code

bwoshea committed:
merged!

--HG--
branch : week-of-code

matthewturk committed:
Merging

--HG--
branch : week-of-code

matthewturk committed:
Looks like PyArray_ENABLEFLAGS is not available until NumPy-1.7.1.

--HG--
branch : week-of-code

matthewturk committed:
Merge

--HG--
branch : week-of-code

jwise77 committed:
Adding ability to read external species fields for PhotonTest.

--HG--
branch : week-of-code

jwise77 committed:
Adding ability to read external species fields for PhotonTest.

--HG--
branch : litemp

jwise77 committed:
Merging

--HG--
branch : week-of-code

jwise77 committed:
Calculate mu directly from data for initial temperature.

--HG--
branch : litemp

jwise77 committed:
Calculate mu directly from data for initial temperature.

--HG--
branch : week-of-code

jwise77 committed:
Turning off ray boundary warning.

--HG--
branch : week-of-code

jwise77 committed:
Place a floor for external density fields.

--HG--
branch : week-of-code

jwise77 committed:
Cleaning up rate solver calls in EvolvePhotons

--HG--
branch : week-of-code

jwise77 committed:
1. Ability to calculate X-ray flux in the optically thin limit (no
support for stars or tree acceleration yet).

2. Mark subgrid marker for rays to NULL for ghost zones outside the
domain for isolated boundary conditions.

--HG--
branch : week-of-code

jwise77 committed:
Extending optically-thin X-rays to stars.

--HG--
branch : week-of-code

jwise77 committed:
Commented out tree-accelerated optically-thin x-rays.

--HG--
branch : week-of-code

jwise77 committed:
Enforcing a density floor for initial density field

--HG--
branch : week-of-code

jwise77 committed:
When using optically-thin X-rays, must solve for chem/energy for all
grids.  Changed looping order for better efficiency.

--HG--
branch : week-of-code

jwise77 committed:
Merging

--HG--
branch : week-of-code

jwise77 committed:
Decreasing critical photon flux for deletion.

--HG--
branch : week-of-code

jwise77 committed:
Fixes to compile with PGI.

--HG--
branch : week-of-code

jwise77 committed:
Merging

--HG--
branch : week-of-code

Commit:
Merging woc heads

--HG--
branch : week-of-code

Commit:
Merging week-of-code into litemp

--HG--
branch : litemp

jwise77 committed:
Putting SSE calls in a USE_SSE define.

--HG--
branch : litemp

jwise77 committed:
Merging

--HG--
branch : litemp

Commit:
Merging

--HG--
branch : litemp

samskillman committed:
First pass at threading PPM, ffts. Also set up a switch so that level 0 and
levels with less than Nprocs*Nthreads do not thread the grid loop.  Finally,
modified the EnzoTiming to only use the master thread for timing.  This is still
a work in progress.

--HG--
branch : litemp

Commit:
Changes to scheduling.  Writing # of cores / MPI in Evtime.  Adding
RaySegments to fields to be removed if RT is turned off.

--HG--
branch : litemp

Commit:
Merge

--HG--
branch : litemp

Commit:
Don't include SN feedback outside the refine region.

--HG--
branch : litemp

samskillman committed:
Modifying some openmp stuff, turning of threading of RH for now.

--HG--
branch : litemp

samskillman committed:
Merging in enzo-bluewaters fork

--HG--
branch : litemp

samskillman committed:
merging woc heads

--HG--
branch : week-of-code

samskillman committed:
Merging in woc to litemp

--HG--
branch : litemp

samskillman committed:
This was merged incorrectly.  Fixes the ReadGhostZones issue again.

--HG--
branch : litemp

jwise77 committed:
Need OPENMP defines to compile without openmp.

--HG--
branch : litemp

jwise77 committed:
Merging litemp branches from my enzo-dev and openmp forks.

--HG--
branch : litemp

jwise77 committed:
Merging latest changes from week-of-code into litemp.

--HG--
branch : litemp

jwise77 committed:
Merging litemp branches

--HG--
branch : litemp

jwise77 committed:
Removing star lifetime restriction for timestep.

--HG--
branch : litemp

jwise77 committed:
Merging in some more changes I forgot on the GT cluster.

--HG--
branch : litemp

jwise77 committed:
Use RadiationFieldRedshift for Compton cooling/heating.

--HG--
branch : litemp

samskillman committed:
Fixing continuation for openmp

--HG--
branch : litemp

samskillman committed:
Threading a few more routines.

--HG--
branch : litemp

samskillman committed:
adding omp master directives to timing

--HG--
branch : litemp

samskillman committed:
removing collapse(3) from utilities that was causing a segfault

--HG--
branch : litemp

samskillman committed:
Modifying collapse statements which led to race conditions, I think.  Also added timing to communication calls.

--HG--
branch : litemp

jwise77 committed:
Adding a new radiative transfer parameter that limits the timestep
based on a sound crossing time on a user-specified level.

--HG--
branch : litemp

Commit:
Re-merging, this time correctly.

--HG--
branch : litemp

Commit:
Needed to include some .defs

--HG--
branch : litemp

jwise77 committed:
Removing parallel region from grid_cic.

--HG--
branch : litemp

jwise77 committed:
Moving 50% timestep increase restriction to high density chemistry.
Making debugging statements thread safe.

--HG--
branch : litemp

jwise77 committed:
Merging.

--HG--
branch : litemp

jwise77 committed:
Minor fixes to the Pop III accretion sphere and chemistry/energy
timestepping.

--HG--
branch : litemp

jwise77 committed:
Merging litemp heads.

--HG--
branch : litemp

Commit:
Merging in an intermediate changeset from the main branch

--HG--
branch : litemp

Commit:
Attempt to merge to tip.

--HG--
branch : litemp

Commit:
Fixing from a not-completely successful merge, with help from Dan on solve_rate_cool.F90.

--HG--
branch : litemp

Commit:
Fixing bugs with grid_cic.F90 conversion, solve_rate_cool, and fixing compilation
when using openmp-no.

--HG--
branch : litemp

Commit:
Removing grid_cic.F

--HG--
branch : litemp

jwise77 committed:
Merging swskillman's openmp changes and merge from the mainline into
my litemp fork.

--HG--
branch : litemp

jwise77 committed:
Fixing merge conflict.

--HG--
branch : litemp

jwise77 committed:
Changing some leftover FORTRAN logicals to LOGIC_PREC.

--HG--
branch : litemp

jwise77 committed:
Reverting some changes that I made in the litemp branch, attempting to
make the energy/chem solver more stable in the presence of radiation.

--HG--
branch : litemp

jwise77 committed:
Removed extra argument in the call to solve_cool.F, causing errors in
sims with only radiative cooling and MultiSpecies == 0.

--HG--
branch : litemp

jwise77 committed:
Merging in Dave's acceleration fix.

--HG--
branch : week-of-code

jwise77 committed:
Merging latest changes from week-of-code into litemp branch.

--HG--
branch : litemp

jwise77 committed:
Don't add NULL subgrid markers from parent grids.  Fixes a segfault.

--HG--
branch : litemp

jwise77 committed:
Expanding the load balancing timers to accommodate multiple timers.
First application will be for ray tracing.

--HG--
branch : litemp

jwise77 committed:
Some things that were overlooked when expanding the load balancing
timers to multiple values.  The Ray_Segment field is still being
corrupted in parallel jobs, which is introduced in the earlier
changeset to avoid interpolation of this field.

--HG--
branch : litemp

jwise77 committed:
RaySegments should not be interpolated and should be zeroed out in
regions with child grids, i.e. rays are only traced on the finest
level.

--HG--
branch : litemp

jwise77 committed:
Adding EnzoTimers to EvolvePhotons and grid::TransportPhotons.

--HG--
branch : litemp

jwise77 committed:
Fixing a long-standing bug where a photon was orphaned when the
MoveToGrid was computed incorrectly.

--HG--
branch : litemp

jwise77 committed:
Adding openmp flags to BW machine file.  Removing cic_deposit omp
pragma since it was causing crashes.

--HG--
branch : litemp

Commit:
Moved cic_deposit omp section to PrepareDensityField

--HG--
branch : cjmp

Commit:
Moved openmp sections to top level loops (iterating over grids)

--HG--
branch : cjmp

Commit:
Added condition for Interpolate to detect if it is running serially

--HG--
branch : cjmp

Commit:
OpenMP implementation of interpolation

--HG--
branch : cjmp

Commit:
Added comment to xeonphi makefile

--HG--
branch : cjmp

Commit:
Cleaning up debug statements, added some more debugging to CommunicationInitialize.C

--HG--
branch : cjmp

Commit:
Patching in last few changes

--HG--
branch : cjmp

Commit:
added the AGN particle files

Commit:
rmoved merge conflicts

Commit:
removed merge conflicts

DeovratPrasad committed:
Active particlee-AGNParticle

DeovratPrasad committed:
test commit

DeovratPrasad committed:
modified how AGN particle is introduced in the domain.

bigtoasty committed:
Cleaned up AGN particle creation routine

bigtoasty committed:
Added fixes that require static analysis

bigtoasty committed:
Particle will now spawn on highest subgrid that matches its position

bigtoasty committed:
many changes

bigtoasty committed:
last few bug fixes

cjllorente827 committed:
merge with cosmo_fix

cjllorente827 committed:
Merging openmp branch

bwoshea commented:
@cjllorente827 could you please (1) update the text of the PR to explain what you're doing with it, and (2) fix the code so it passes the tests?  Thank you!

jwise77 committed:
Fixing merge conflicts. Compiles now. Adding GEMINI.md

jwise77 committed:
Fixing some more merge errors. Compiles.

jwise77 committed:
Merge branch 'main' into openmp_merge

jwise77 committed:
Fix OpenMP data races and nowait drift in particle collection/move routines

- Add missing array reductions for StarsToMove and APNumberToMove in CommunicationCollectParticles.C and CommunicationTransferSubgridParticles.C.
- Restrict parallel block in CommunicationTransferSubgridParticles.C to exclude active particle copy loop.
- Remove nowait clauses from inner work-sharing loops nested inside sequential loops in Grid_CollectParticles.C, Grid_MoveAllParticles.C, and Grid_CommunicationSendParticles.C to ensure compliant work-sharing ordering.

jwise77 committed:
Merge branch 'main' into openmp_merge

jwise77 committed:
Revert AGN particle commits that were accidentally included in the openmp branch

- Revert "Particle will now spawn on highest subgrid that matches its position"
  - This reverts commit d27a9b2785929ca66576d14404df116df11a7711.
- Revert "rmoved merge conflicts"
  - This reverts commit 48c06a701bbc91e50309d42f8e8af1b40a985bba,
    reversing changes made to 277c0a8e4522a4e71e044bdc0897ad58b0b07322.
- Revert "Bugfix for ActiveParticleFinalize function call"
  - This reverts commit 815bd8137012ff783e3ec4109eedfd22a50d1c37.
- Revert "added the AGN particle files"
  - This reverts commit 277c0a8e4522a4e71e044bdc0897ad58b0b07322.

jwise77 committed:
Fix bugs in openmp_merge branch including dep_grid_cic signature mismatch, FindSubgrids pointer bug, and EvolveLevel loop nesting

jwise77 committed:
Fix race conditions in IdentifyNewSubgridsBySignature by removing static GridEnds array and restore thread_grid_loop logic

jwise77 committed:
Parallelize chemistry and cooling grid loop in EvolveLevel.C to improve multi-threaded scalability

jwise77 committed:
Fix MPI truncate errors in hybrid MPI+OpenMP by only parallelizing density field loops when NumberOfProcessors == 1

jwise77 committed:
Parallelize Green's function multiplication in ComputePotentialFieldLevelZero.C

jwise77 committed:
Fix segfault in TransportPhotonPackages by declaring MoveToGrid and ThreadedMoveList inside the parallel block to make them thread-local

jwise77 committed:
Fix duplicate star particle crashes by making TransferSubgridStars and TransferSubgridActiveParticles thread-safe using shared counters and omp critical sections

jwise77 committed:
Update Sphinx documentation to describe OpenMP compilation, threaded run execution, hybrid OpenMP+MPI configurations, and code threading structure
